### PR TITLE
Add MapType as JSON-compatible

### PR DESCRIPTION
**Pull request by srowen**
## Description of changes:

`dataframe_to_mds` does not directly support the Spark Map type as there is no direct analog in streaming (?)
However, the dataframe_to_mds will accept "unsupported" Spark types if they are JSON-compatible, and if the user specifies type 'json' for the column.

However the logic of `is_json_compatible` doesn't include `MapType`, when a map is conceptually fine as a dict, and Spark uses simple dicts to represent such maps in pandas DFs. As long as the map's key and value types are JSON compatible, it should be considered JSON compatible.

With this change, it's possible to have columns including a map type successfully convert to JSON in an MDS file if the given type for the column is specified as 'json'.

### General
- [X] I have read the [contributor guidelines](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md)
- [ ] This is a documentation change or typo fix. If so, skip the rest of this checklist.
- [X] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the MosaicML team.
- [X] I have updated any necessary documentation, including [README](https://github.com/mosaicml/streaming/blob/main/README.md) and [API docs](https://github.com/mosaicml/streaming/tree/main/docs) (if appropriate).

### Tests
- [X] I ran `pre-commit` on my change. (check out the `pre-commit` section of [prerequisites](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#prerequisites))
- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate).
- [ ] I ran the tests locally to make sure it pass. (check out [testing](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#submitting-a-contribution))
- [ ] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes.

### CODE REVIEW

File: tests/base/converters/test_dataframe_to_mds.py [hunk 1]
```diff
@@ -418,13 +418,18 @@ def test_is_json_compatible(self):
                     ]), True), True)
         ])
 
-        valid_schemas = [message_schema, prompt_response_schema, combined_schema]
+        schema_with_string_map_keys = StructType(
```
Comment (snarayan21) [inline tests/base/converters/test_dataframe_to_mds.py:421]:
```suggestion
        string_map_keys_schema = StructType(
```
just to make variable naming the same as the other schemas

File: tests/base/converters/test_dataframe_to_mds.py [hunk 2]
```diff
@@ -418,13 +418,18 @@ def test_is_json_compatible(self):
...
+            [StructField('map_field', MapType(StringType(), StringType()), nullable=True)])
+
+        valid_schemas = [
+            message_schema, prompt_response_schema, combined_schema, schema_with_string_map_keys
```
Comment (snarayan21) [inline tests/base/converters/test_dataframe_to_mds.py:425]:
```suggestion
            message_schema, prompt_response_schema, combined_schema, string_map_keys_schema
```
see above

File: tests/base/converters/test_dataframe_to_mds.py [hunk 3]
```diff
@@ -418,13 +418,18 @@ def test_is_json_compatible(self):
...
+        ]
 
         schema_with_binary = StructType([StructField('data', BinaryType(), nullable=True)])
 
         # Schema with MapType having non-string keys
         schema_with_non_string_map_keys = StructType(
-            [StructField('map_field', MapType(IntegerType(), StringType()), nullable=True)])
+            [StructField('map_field', MapType(BinaryType(), StringType()), nullable=True)])
```
Comment (snarayan21) [inline tests/base/converters/test_dataframe_to_mds.py:432]: why change to BinaryType, just wondering?

**Review by snarayan21 — COMMENTED** [review]
minor nits.

If possible can you add a simple unit test?

File: tests/base/converters/test_dataframe_to_mds.py [hunk 4]
```diff
@@ -418,13 +418,18 @@ def test_is_json_compatible(self):
...
+        valid_schemas = [
+            message_schema, prompt_response_schema, combined_schema, string_map_keys_schema
+        ]
 
         schema_with_binary = StructType([StructField('data', BinaryType(), nullable=True)])
 
         # Schema with MapType having non-string keys
         schema_with_non_string_map_keys = StructType(
-            [StructField('map_field', MapType(IntegerType(), StringType()), nullable=True)])
+            [StructField('map_field', MapType(BinaryType(), StringType()), nullable=True)])
```
Comment (snarayan21) [inline tests/base/converters/test_dataframe_to_mds.py:432]: nit: above this line, can you change `schema_with_non_string_map_keys` variable name to `non_string_map_keys_schema` as well?

**Review by snarayan21 — APPROVED** [review]
minor nit: variable name change. otherwise lgtm

**Comment by srowen** [comment]
@snarayan21 said I could merge after that change, so here goes!

**Pull request merged**